crates/bevy_pbr/src/material_bind_groups.rs

```diff
@@ -235,15 +296,156 @@ impl<M> MaterialBindGroup<M>
 where
     M: Material,
 {
-    /// Returns a new bind group.
-    fn new(bindless_enabled: bool) -> MaterialBindGroup<M> {
-        let count = if !bindless_enabled {
-            1
+    /// Creates a new material bind group.
+    fn new(bindless: bool) -> MaterialBindGroup<M> {
+        if bindless {
+            MaterialBindGroup::Bindless(MaterialBindlessBindGroup::new())
         } else {
-            M::BINDLESS_SLOT_COUNT.unwrap_or(1)
-        };
+            MaterialBindGroup::NonBindless(MaterialNonBindlessBindGroup::new())
+        }
+    }
+
+    /// Allocates a new binding slot and returns its ID.
+    fn allocate(&mut self) -> MaterialBindGroupSlot {
+        match *self {
+            MaterialBindGroup::Bindless(ref mut material_bindless_bind_group) => {
+                material_bindless_bind_group.allocate()
+            }
+            MaterialBindGroup::NonBindless(ref mut material_non_bindless_bind_group) => {
+                material_non_bindless_bind_group.allocate()
+            }
+        }
+    }
+
+    /// Assigns an unprepared bind group to the group and slot specified in the
+    /// [`MaterialBindingId`].
+    fn init(
+        &mut self,
+        render_device: &RenderDevice,
+        bind_group_layout: &BindGroupLayout,
+        slot: MaterialBindGroupSlot,
+        unprepared_bind_group: UnpreparedBindGroup<M::Data>,
+    ) {
+        match *self {
+            MaterialBindGroup::Bindless(ref mut material_bindless_bind_group) => {
+                material_bindless_bind_group.init(
+                    render_device,
+                    bind_group_layout,
+                    slot,
+                    unprepared_bind_group,
+                );
+            }
+            MaterialBindGroup::NonBindless(ref mut material_non_bindless_bind_group) => {
+                material_non_bindless_bind_group.init(
+                    render_device,
+                    bind_group_layout,
+                    slot,
+                    unprepared_bind_group,
+                );
+            }
+        }
+    }
+
+    /// Fills a slot directly with a custom bind group.
+    ///
+    /// This is only a meaningful operation for non-bindless bind groups. It's
+    /// rarely used, but see the `texture_binding_array` example for an example
+    /// demonstrating how this feature might see use in practice.
+    fn init_custom(&mut self, bind_group: BindGroup, extra_data: M::Data) {
+        match *self {
+            MaterialBindGroup::Bindless(_) => {
+                error!("Custom bind groups aren't supported in bindless mode");
```

Comment:
Dumb question (not asking for any changes), does limiting custom bind groups to non-bindless mean eg. custom materials (thinking about either an extended material, or maybe something like a full `StandardMaterial` toon shader replacement?) won't benefit from bindless currently?

Reply:
Bindless works for any material as long as you write `#[bindless]` on it; there's nothing specific to `StandardMaterial`. (Actually, `StandardMaterial` doesn't benefit from bindless yet; that's #16644.)

The restriction you quoted only applies in very unusual circumstances, basically only in the `custom_shader_instancing` example.